packages/foam-vscode/src/features/preview/wikilink-embed.ts

```diff
@@ -47,25 +47,20 @@ export const markdownItWikilinkEmbed = (
         let content = `Embed for [[${wikilink}]]`;
         switch (includedNote.type) {
           case 'note': {
-            let noteText = readFileSync(includedNote.uri.toFsPath()).toString();
-            const section = Resource.findSection(
+            const noteStyle = getFoamVsCodeConfig(
+              CONFIG_EMBED_NOTE_IN_CONTAINER
+            )
+              ? 'card'
+              : 'inline';
+            const noteEmbedder = new NoteEmbedder(
```

Comment:
I wonder if `NoteEmbedder` helps encapsulate a behavior, and compare that with the additional "complexity" of the indirection - what do you think?

Reply:
A little confused by "encapsulate a behavior". Is that a type of implementation pattern?

Reply:
no no, what I mean is basically that currently in `NoteEmbedder` lives the code that was in this function (that's what I mean by "encapsulating": we don't see the actual code and rely on this object to "do the work").
This indirection removes complexity by removing some code here (and putting it in `NoteEmbedder`) but also adds complexity because understanding the code requires an extra step (that is, check the implementation of `NoteEmbedder`).
My original comment was whether what's gained offsets what's lost, I wonder if we could in fact remove the `NoteEmbedder` and simply have the code here, wdyt?

Reply:
Gotcha, I see what you mean. That's a good point. For https://github.com/foambubble/foam/pull/1273/commits/e05c1c282e59e995e8c707f8787de8157e73eb05 I changed the interfaces to function types and replaced all the classes with functions. Is this what you had in mind? 

I feel like the class-based approach would be preferable after we add more variations, so maybe we can go with the functional approach and reconsider it as an option in the future? I say that because with the functional approach, it seems dependencies get added for all formatters and extractors. Like if extractor A requires `parser: ResourceParser`, but extractor B doesn't,  the type still needs to include it, and extractor B has to have it in its signature:
```
export type EmbedNoteExtractor = (
  note: Resource,
  parser: ResourceParser,
) => string;

function extractorA(note: Resource, parser: resourceParser) {}
function extractorB(note: Resource, parser: resourceParser) {} // doesn't need the parser, but generateNoteEmbedding() expects a type EmbedNoteExtractor 
```

Reply:
How about moving the two lines of `generateNoteEmbedding` in the caller and remove the fn altogether?

Your comment about the dependencies is correct. And in some of my functions this situation happens. For example in the feature activate function I used as example before, sometimes the `Foam` object is not required, but I pass it anyways.
Depending on the number of dependencies and how related they are to the method, I am ok with the compromise.

In this specific case of the extractor and the formatter, so far the parameters being passed IMO read well with the function signature, so I don't see yellow/red flags. 
This is something that can be reassesed in the future shall the need arise. 

---

As an aside, the functional way of mimicking the constructor would be using a factory, e.g.

```
function createExtractor(extraDep: any) : (note: Resource, parser: resourceParser) => string {
  return (note: Resource, parser: resourceParser) => {
    // do something with all extraDep, note and parser
  }
}

extractorC = createExtractor(extra)

extractorC(note, parser)
```

Reply:
Great suggestion, done! Also really appreciate the snippet